File: backend/app/agents/hotspot_agent.py
from __future__ import annotations

from app.agents.agentscope_runtime import run_structured_agent_sync
from app.agents.common import HOT_KEYWORDS, extract_keywords, infer_priority, infer_trend, normalize_whitespace
from app.agents.context import AgentExecutionContext
from app.agents.structured_models import HotspotStructuredOutput


def _build_hotspot_user_prompt(context: AgentExecutionContext) -> str:
    news = context.news
    return "\n".join(
        [
            "请判断下面这条 AI 新闻是否值得继续作为热点处理，只输出结构化结果，不要额外解释。",
            f"标题：{news.title}",
            f"来源：{news.source}",
            f"分类：{news.category or '未分类'}",
            f"热度分值：{news.hot_score}",
            f"语言：{news.language}",
            f"发布时间：{news.publish_time.isoformat() if news.publish_time else '未知'}",
            f"摘要：{news.summary or '无'}",
            "正文：",
            news.content or "",
        ]
    )


def run_hotspot_agent(context: AgentExecutionContext) -> dict[str, object]:
    runtime = context.runtime
    if runtime is not None:
        agent_result = run_structured_agent_sync(
            runtime=runtime,
            agent_name="HotspotAgent",
            system_prompt=context.prompt_bundle.get("agent_hotspot_prompt", ""),
            user_prompt=_build_hotspot_user_prompt(context),
            structured_model=HotspotStructuredOutput,
        )
        if agent_result is not None:
            return {
                **agent_result,
                "model_name": context.model_name,
                "prompt_version": context.prompt_version,
            }

    news = context.news
    text = f"{news.title} {news.content or ''}"
    keywords = extract_keywords(text)
    hot_score = int(news.hot_score or 0)
    if any(keyword.lower() in text.lower() for keyword in HOT_KEYWORDS):
        hot_score += 8
    if news.source and any(marker in news.source.lower() for marker in ("openai", "anthropic", "deepseek", "qwen", "modelscope")):
        hot_score += 5

    hot_score = min(hot_score, 100)
    is_hot = hot_score >= context.config["hot_threshold"] or news.status in {"FILTERED", "SCRIPT_READY"}
    reason_parts = [
        f"热度评分 {hot_score}",
        f"来源 {news.source}",
        f"关键词 {', '.join(keywords[:3])}" if keywords else "未命中显著关键词",
    ]
    if news.summary:
        reason_parts.append(f"已有摘要 {normalize_whitespace(news.summary)[:60]}")

    return {
        "is_hot": is_hot,
        "score": hot_score,
        "priority": infer_priority(hot_score),
        "trend": infer_trend(hot_score, title=news.title, content=news.content or ""),
        "reason": "；".join(reason_parts),
        "keywords": keywords[:5],
    }

File: backend-ts/src/agents/hotspot.agent.ts
/**
 * HotspotAgent - Hotspot judgment agent
 * Determines whether input news is a high-value AI hotspot
 *
 * Pattern: AI-first rule-fallback
 */

import { Injectable, Logger } from '@nestjs/common';
import { z } from 'zod';

import { AgentScopeRuntimeService } from './agents.runtime';
import type { AgentExecutionContext, HotspotOutput } from './agents.types';
import {
  HOT_KEYWORDS,
  extractKeywords,
  inferPriority,
  inferTrend,
  normalizeWhitespace,
} from './agents.common';

// ===== Zod Structured Output Schema =====

export const HotspotSchema = z.object({
  is_hot: z.boolean(),
  score: z.number().min(0).max(100),
  priority: z.enum(['HIGH', 'MEDIUM', 'LOW']),
  trend: z.enum(['RISING', 'WATCH', 'STABLE']),
  reason: z.string(),
  keywords: z.array(z.string()).default([]),
});

@Injectable()
export class HotspotAgent {
  private readonly logger = new Logger(HotspotAgent.name);

  constructor(private readonly runtime: AgentScopeRuntimeService) {}

  /** Build user prompt */
  private buildUserPrompt(ctx: AgentExecutionContext): string {
    const news = ctx.news;
    return [
      'Determine whether this AI news is worth processing as a hotspot. Output structured JSON only.',
      `Title: ${news.title}`,
      `Source: ${news.source}`,
      `Category: ${news.category || 'N/A'}`,
      `Hot Score: ${news.hot_score}`,
      `Language: ${news.language}`,
      `Published: ${news.publish_time?.toISOString() || 'unknown'}`,
      `Summary: ${news.summary || 'N/A'}`,
      'Content:',
      news.content || '',
    ].join('\n');
  }

  /**
   * Execute hotspot judgment
   * Returns HotspotOutput, never throws (fallback guaranteed)
   */
  async run(ctx: AgentExecutionContext): Promise<HotspotOutput> {
    // Try AI call first
    const aiResult = await this.runtime.callStructured<HotspotOutput>({
      agentName: 'HotspotAgent',
      systemPrompt: ctx.prompt_bundle.agent_hotspot_prompt || '',
      userPrompt: this.buildUserPrompt(ctx),
      schema: HotspotSchema,
    });

    if (aiResult) {
      this.logger.debug(`HotspotAgent AI result: is_hot=${aiResult.is_hot}, score=${aiResult.score}`);
      return aiResult;
    }

    // Rule-based fallback
    return this.fallback(ctx);
  }

  /** Rule fallback logic */
  private fallback(ctx: AgentExecutionContext): HotspotOutput {
    const news = ctx.news;
    const text = `${news.title} ${news.content || ''}`.toLowerCase();
    const keywords = extractKeywords(text);
    let hotScore = news.hot_score || 0;

    // Hot keyword bonus
    if (HOT_KEYWORDS.some((kw) => text.includes(kw))) {
      hotScore += 8;
    }
    // Well-known source bonus
    if (
      news.source &&
      ['openai', 'anthropic', 'deepseek', 'qwen', 'modelscope'].some((marker) =>
        news.source!.toLowerCase().includes(marker),
      )
    ) {
      hotScore += 5;
    }

    hotScore = Math.min(hotScore, 100);
    const threshold = (ctx.config.hot_threshold as number) ?? 35;
    const isHot =
      hotScore >= threshold ||
      news.status === 'FILTERED' ||
      news.status === 'SCRIPT_READY';

    const reasonParts: string[] = [`Hot score ${hotScore}`, `Source: ${news.source}`];
    if (keywords.length > 0) {
      reasonParts.push(`Keywords: ${keywords.slice(0, 3).join(', ')}`);
    } else {
      reasonParts.push('No significant keywords matched');
    }
    if (news.summary) {
      reasonParts.push(`Summary: ${normalizeWhitespace(news.summary).slice(0, 60)}`);
    }

    return {
      is_hot: isHot,
      score: hotScore,
      priority: inferPriority(hotScore),
      trend: inferTrend(hotScore, news.title, news.content || ''),
      reason: reasonParts.join('; '),
      keywords: keywords.slice(0, 5),
    };
  }
}
